### PR TITLE
6-auth-section

### PR DESCRIPTION
**Pull request by toshmukhamedov**
**Description:**

This pull request introduces the implementation of two new APIs: POST /auth and POST /auth/verification/code. These APIs enhance our authentication system by providing endpoints for user authentication and email verification code generation, respectively.

**Changes:**

1. **POST /auth Endpoint:**
   - Implemented the POST /auth endpoint to handle user authentication.
   - Added route handler to process user credentials and generate authentication tokens.
   - Implemented validation checks for user credentials.
   - Integrated JWT (JSON Web Tokens) for secure token generation and authentication.

2. **POST /auth/verification/code Endpoint:**
   - Implemented the POST /auth/verification/code endpoint for generating email verification codes.
   - Added route handler to generate and send verification codes to users via email.
   - Implemented logic to validate user identity before generating the code.
   - Enhanced security measures to prevent abuse and unauthorized access.

**Testing:**

- Conducted thorough testing of both APIs using unit tests and integration tests.
- Verified functionality, including successful authentication and email verification code generation.
- Ensured error handling and edge cases are handled appropriately.
- Tested API responses against expected outcomes to ensure consistency and correctness.

**Documentation:**

- Updated API documentation to include details of the newly added endpoints.
- Documented request and response formats for both /auth and /auth/verification/code APIs.
- Included usage examples and guidelines for developers integrating with these endpoints.

**Additional Notes:**

- This implementation aligns with our project's authentication requirements and enhances user security with email verification.
- Feedback and suggestions for improvement are welcome and appreciated.

**Reviewer(s):**

- @Xayrulloh

### CODE REVIEW

File: .github/workflows/pull_request.yml
Comment (Xayrulloh) [inline .github/workflows/pull_request.yml:1]: nma qiladi bu

File: .gitignore [hunk 1]
```diff
@@ -14,4 +14,4 @@ dist
 .env
 
 # incremental compilation
-.tsbuildinfo
\ No newline at end of file
+.tsbuildinfo
```
Comment (Xayrulloh) [inline .gitignore:17]: what happened

File: src/config/env.ts [hunk 2]
```diff
@@ -14,7 +21,7 @@ const result = schema.safeParse(process.env);
 if (!result.success) {
     console.error(result.error.issues);
     console.error("Some Environment variables are missing. Exiting...");
-    process.exit(1);
+    process.exit(ProcessCode.ERROR);
```
Comment (Xayrulloh) [inline src/config/env.ts:24]: nma bo'votti bo'tta 🤔

File: src/db/schema/users.ts [hunk 3]
```diff
@@ -3,9 +3,10 @@ import { pgTable, serial, varchar } from "drizzle-orm/pg-core";
 
 export const users = pgTable("users", {
     id: serial("id").primaryKey(),
-    name: varchar("name", { length: 64 }).notNull(),
+    name: varchar("name", { length: 64 }),
```
Comment (Xayrulloh) [inline src/db/schema/users.ts:6]: ismi bo'masligi mumkinmi wunda

File: src/db/schema/verification-codes.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { base } from "@utils/base.schema.ts";
+import { timestamp, integer, pgTable, serial, varchar } from "drizzle-orm/pg-core";
+import { users } from "./users.ts";
+
+export const verificationCodes = pgTable("verification_codes", {
+    id: serial("id").primaryKey(),
```
Comment (Xayrulloh) [inline src/db/schema/verification-codes.ts:6]: ...base icida id ham boridi wekl

File: src/integrations/email.service.ts
Comment (Xayrulloh) [inline src/integrations/email.service.ts:1]: wuni 1ta gaplawvose zo'r bo'lardi, cunvolardm

File: src/integrations/templates/verification-code.html
Comment (Xayrulloh) [inline src/integrations/templates/verification-code.html:1]: nice

File: src/main.ts [hunk 5]
```diff
@@ -1,23 +1,43 @@
...
+});
 
-fastify.get("/", async (request, reply) => {
-    reply.send({ hello: "world" });
+// INFO: Swagger
+registerSwagger(fastify);
+
+// Declare a route
+fastify.register(routes, {
+    prefix: "v1",
```
Comment (Xayrulloh) [inline src/main.ts:34]: nma qiladi bu

File: src/main.ts [hunk 6]
```diff
@@ -1,23 +1,43 @@
...
 
-// Declare a route
-fastify.register(routes);
+// INFO: Add schema validator and serializer
+// fastify.setValidatorCompiler(validatorCompiler);
+// fastify.setSerializerCompiler(serializerCompiler);
+
+await fastify.register(import("@fastify/middie"));
+await fastify.register(cors);
+await fastify.register(helmet);
```
Comment (Xayrulloh) [inline src/main.ts:21]: gj

File: src/routes/index.ts [hunk 7]
```diff
@@ -1,14 +1,20 @@
+import { authHookHandler } from "@hooks/auth.hook.ts";
 import type { FastifyInstance } from "fastify";
 import authRoutes from "./auth/auth.route.ts";
 import groupRoutes from "./group.route.ts";
 import listRoutes from "./list.route.ts";
 import userRoutes from "./user.route.ts";
 
 const routes = async (fastify: FastifyInstance) => {
+    // hooks
+    fastify.addHook("preHandler", authHookHandler);
```
Comment (Xayrulloh) [inline src/routes/index.ts:10]: nice

File: src/utils/enums.ts [hunk 8]
```diff
@@ -1,10 +1,74 @@
...
+    Actual = "actual",
+    Future = "future",
+}
+
+export const enum ProcessCode {
+  SUCCESS = 0,
+  ERROR = 1
+}
+
+export const enum HttpStatus {
```
Comment (Xayrulloh) [inline src/utils/enums.ts:17]: damn, o'zida yo'g'akammi. Sizam tekis nestjs digi httpStatus dan copib keldismi dimande 😂

File: src/utils/helpers.spec.ts
Comment (Xayrulloh) [inline src/utils/helpers.spec.ts:1]: erinmagan odamakansiz 😂

**Review by Xayrulloh — CHANGES_REQUESTED** [review]

File: .github/workflows/pull_request.yml
Comment (toshmukhamedov) [reply .github/workflows/pull_request.yml:1]: test run qiladi

File: .gitignore (hunk 1 above)
Comment (toshmukhamedov) [reply .gitignore:17]: incremental compilation

File: src/config/env.ts (hunk 2 above)
Comment (toshmukhamedov) [reply src/config/env.ts:24]: app to'xtatilvoti

File: src/db/schema/users.ts (hunk 3 above)
Comment (toshmukhamedov) [reply src/db/schema/users.ts:6]: oh no, I've to set extracted name from email
I'll fix

File: src/db/schema/verification-codes.ts (hunk 4 above)
Comment (toshmukhamedov) [reply src/db/schema/verification-codes.ts:6]: good point
I'll add it to `base`

File: src/utils/enums.ts (hunk 8 above)
Comment (toshmukhamedov) [reply src/utils/enums.ts:17]: yes

File: src/main.ts (hunk 6 above)
Comment (toshmukhamedov) [reply src/main.ts:21]: actually we don't need `middie` I'll remove it

File: src/main.ts (hunk 5 above)
Comment (toshmukhamedov) [reply src/main.ts:34]: API https://api.planit.uz/v1/**
i.e `https://api.planit.uz/v1/auth`, `https://api.planit.uz/v1/auth/verification/code`

File: src/integrations/email.service.ts
Comment (toshmukhamedov) [reply src/integrations/email.service.ts:1]: when?

File: src/db/schema/verification-codes.ts (hunk 4 above)
Comment (toshmukhamedov) [reply src/db/schema/verification-codes.ts:6]: let's remove `base` and define each table indepently?

Comment (toshmukhamedov) [reply src/db/schema/verification-codes.ts:6]: because it's not suitable for all cases

File: .github/workflows/pull_request.yml
Comment (Xayrulloh) [reply .github/workflows/pull_request.yml:1]: nmani testini

File: .gitignore (hunk 1 above)
Comment (Xayrulloh) [reply .gitignore:17]: cummadimu nu mayli

File: src/config/env.ts (hunk 2 above)
Comment (Xayrulloh) [reply src/config/env.ts:24]: nmaga 1 o'rniga .ERROR ili o'wani icida 1 bormi 🤔

File: src/db/schema/users.ts (hunk 3 above)
Comment (Xayrulloh) [reply src/db/schema/users.ts:6]: emaildan danniy uhlatvovosmi wunda 🤔

File: src/db/schema/verification-codes.ts (hunk 4 above)
Comment (Xayrulloh) [reply src/db/schema/verification-codes.ts:6]: > let's remove `base` and define each table indepently?

nope

Comment (Xayrulloh) [reply src/db/schema/verification-codes.ts:6]: > because it's not suitable for all cases

okay

File: .github/workflows/pull_request.yml
Comment (toshmukhamedov) [reply .github/workflows/pull_request.yml:1]: umuman qanaqa test yozilgan bo'lsa hammasini, mrdan keyin buzilib qolmasligini oldini olish uchun

File: src/config/env.ts (hunk 2 above)
Comment (toshmukhamedov) [reply src/config/env.ts:24]: yes it's just wrapped version

File: src/integrations/email.service.ts
Comment (Xayrulloh) [reply src/integrations/email.service.ts:1]: kecasiga vaqtis bomi, man hozi tut teriwim kere, keyn svet o'rnatiwim aliexpressdan kegan ya'ni oddiy lampockani burab qo'yiwi etvotilami dib o'ylamasligiscun etib qo'yvomman 😁 , keyn jiyanlarim uyg'onda keregov wu bn opam ketganlaridan keyn co'm co'm qibo'lib namozdan keyn vaqtim bo'larkan

File: src/main.ts (hunk 6 above)
Comment (Xayrulloh) [reply src/main.ts:21]: actually i wrote it for helmet. middie nma qvokanini daje bilganimam yo'q, u nma narsa o'zi

File: .github/workflows/pull_request.yml
Comment (Xayrulloh) [reply .github/workflows/pull_request.yml:1]: haaa

File: src/main.ts (hunk 6 above)
Comment (toshmukhamedov) [reply src/main.ts:21]: fastify/middie: Middleware engine for Fastify

File: src/db/schema/users.ts (hunk 3 above)
Comment (toshmukhamedov) [reply src/db/schema/users.ts:6]: nope I just split by @ and get first element, john@gmail.com -> john

Comment (toshmukhamedov) [reply src/db/schema/users.ts:6]: the reason is make registration quicker

Comment (toshmukhamedov) [reply src/db/schema/users.ts:6]: do you have another idea?

Comment (Xayrulloh) [reply src/db/schema/users.ts:6]: good idea but we should let them change their name after all

**Review by Xayrulloh — APPROVED** [review]
